### PR TITLE
gh-92228: disable the compiler's 'small exit block inlining' optimization for blocks that have a line number

### PR DESCRIPTION
**Pull request by iritkatriel**
Inlining of code that corresponds to source code lines, can make it hard to distinguish later between code which is only reachable from except handlers, and that which is reachable in normal control flow. This caused problems with the debugger's jump feature.

This PR turns off the inlining optimisation for code which has line numbers. We still inline things like the implicit "return None". 

* Issue: gh-92228

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -8981,6 +8981,7 @@ jump_thread(struct instr *inst, struct instr *target, int opcode)
 }
 
 /* Maximum size of basic block that should be copied in optimizer */
+/* Set to 0 to disable to optimisation - see gh-92228 */
```
Comment (iritkatriel) [inline Python/compile.c]:
```suggestion
```

**Comment by tiran** [comment]
Does the PR need yet another bytecode magic bump or can we piggy-back on the magic bump that was merged yesterday?

**Comment by iritkatriel** [comment]
> Does the PR need yet another bytecode magic bump or can we piggy-back on the magic bump that was merged yesterday?

I think it's safe to piggyback. In any case, we didn't change the set of opcodes so I think the old code would normally still work (unless you ask the compiler to jump over an except block to a small exit block).

File: Misc/NEWS.d/next/Core and Builtins/2022-07-06-14-02-26.gh-issue-92228.44Cbly.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Disable the compiler's inline-small-exit-blocks optimization for exit blocks that are associated with source code lines. This fixes a bug where the debugger cannot tell where an exception handler starts and the following code block begins.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2022-07-06-14-02-26.gh-issue-92228.44Cbly.rst]:
```suggestion
Disable the compiler's inline-small-exit-blocks optimization for exit blocks that are associated with source code lines. This fixes a bug where the debugger cannot tell where an exception handler ends and the following code block begins.
```

**Review by tiran — APPROVED** [review]
Tests are passing, benchmark shows no impact, C code changes look fine, test changes are plausible. I trust you know what you are doing, too. :+1:

**Pull request merged**